### ISSUE TITLE
inconsistency in torch.Tensor.scatter on GPU and CPU

### ISSUE THREAD

**Issue by beanduan22** [pytorch/pytorch#114085]
### 🐛 Describe the bug
```python
import torch
import torchvision
import torchvision.transforms as transforms
import torch.nn as nn
import os
import torch.nn.functional as F
gpu_device = torch.device("cuda")
cpu_device = torch.device("cpu")
class PreprocessAndCalculateModel(nn.Module):
    def __init__(self):
        super().__init__()
    def forward(self, x,y,z):
        out = torch.Tensor.scatter(x, 0, y, z)
        return out

real_inputs =  torch.tensor([1, 2, 3, 4, 5, 6, 7])
real_inputs2 = torch.tensor([0, 1, 1, 2])
real_inputs3 = torch.tensor([12, 14, 16, 18, 20])
model = PreprocessAndCalculateModel()
x = real_inputs
output_gpu = model.to(gpu_device)(real_inputs.cuda(),real_inputs2.cuda(),real_inputs3.cuda())
output_cpu = model.to(cpu_device)(real_inputs.cpu(),real_inputs2.cpu(),real_inputs3.cpu())
print(output_gpu)
print(output_cpu)
```
### Versions
```
tensor([12, 14, 18,  4,  5,  6,  7], device='cuda:0')
tensor([12, 16, 18,  4,  5,  6,  7])
```

**Comment by mingfeima** [comment]
@yanbing-j wow, this one sees to be a bug, could you please help with it ?

**Comment by yanbing-j** [comment]
Sure, will do.

**Comment by ptrblck** [comment]
From the [docs](https://pytorch.org/docs/stable/generated/torch.Tensor.scatter_.html#torch.Tensor.scatter_):
> When indices are not unique, the behavior is non-deterministic (one of the values from src will be picked arbitrarily) and the gradient will be incorrect (it will be propagated to all locations in the source that correspond to the same index)!

Based on the source code you are assigning `14` and `16` to index `1`, thus I would claim this is expected behavior. 
If you want to reduce the `src` tensor, use `.scatter_reduce_`.

**Comment by yanbing-j** [comment]
Yes, I agree with @ptrblck , the behavior is non-deterministic. It depends on which value from src is picked. The inconsistency in torch.Tensor.scatter between CPU and GPU is expected.
